mono/mini/mini-llvm.c

```diff
@@ -6335,8 +6338,8 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 					g_assert (values [ins->sreg1]);
 					LLVMBuildStore (builder, convert (ctx, values [ins->sreg1], type_to_llvm_type (ctx, t)), addresses [ins->sreg1]);
 					addresses [ins->dreg] = addresses [ins->sreg1];
-				} else if (values [ins->sreg1] == addresses [ins->sreg1]) {
-					/* LLVMArgVtypeByRef, have to make a copy */
+				} else if (ainfo->storage == LLVMArgVtypeAddr || values [ins->sreg1] == addresses [ins->sreg1]) {
```

Comment:
`LLVMArgVtypeAddr` is only ever used on amd64 and mostly serves the same role as `LLVMArgVtypeByRef`, except for this explicit check and for its use in `process_call`. Is there any harm in dropping `LLVMArgVtypeAddr`, and adding an test here for `LLVMArgVtypeByRef` instead? This would then also cover arm, armv7k, arm64, which also have native ABIs that sometimes pass aggregates via an indirection:

https://github.com/mono/mono/blob/2d06126dc97d09749dae7722fc1a8da44f60e129/mono/mini/mini-arm.c#L1516-L1527

https://github.com/mono/mono/blob/2d06126dc97d09749dae7722fc1a8da44f60e129/mono/mini/mini-arm.c#L2372-L2375

https://github.com/mono/mono/blob/2d06126dc97d09749dae7722fc1a8da44f60e129/mono/mini/mini-arm64.c#L1200-L1204

https://github.com/mono/mono/blob/2d06126dc97d09749dae7722fc1a8da44f60e129/mono/mini/mini-arm64.c#L2435-L2438

Reply:
Tried to avoid hitting cases where OP_LLVM_OUTARG_VT is used but not asking for a valuetype copy and pass by reference of the copy (due to ABI) semantics, so the idea was to make a distinction between the two scenarios all way up to the place where we initially identify that this is coming from the use of ArgValuetypeAddrInIReg/ArgValuetypeAddrOnStack. The other alternative is to add something to the arguments passed along with LLVMArgVtypeByRef, but I don't think that is as clear as having an explicit storage type, the same as we have on mono regular JIT AOT. If the same behavior is needed on other CPU architectures (and that also suffers from the same issue), I believe it is better to use the new storage type for that specific mapping and leave the behavior of ArgVtypeByRef as is (unless it sole purpose was to model this specific case, triggering a valuetype copy and pass that byref according to ABI and not just pass a valuetype byref) . The only place where this bug has been identified so far is on Windows x64 (amd64) and that's where the current fix is applied. If other platforms might be a victim for the same issue, we could leverage the same fix there and make the test triggering the issue more generic.

Reply:
@imhameed ^^^^?

Reply:
@imhameed, thanks! Then I guess this PR is good to go, possible for you to approve it?